src/socket_base.cpp

```diff
@@ -1421,6 +1421,32 @@ void zmq::socket_base_t::process_term_endpoint (std::string *endpoint_)
     delete endpoint_;
 }
 
+void zmq::socket_base_t::process_pipe_stats_publish (
+  uint64_t outbound_queue_count_,
+  uint64_t inbound_queue_count_,
+  endpoint_uri_pair_t *endpoint_pair_)
+{
+    event (*endpoint_pair_, outbound_queue_count_, ZMQ_EVENT_PIPES_STATS);
```

Comment:
TODO: refactor the event function to allow multiple values, so that a single event can be used

Reply:
Another question here: by having the event raised from the application thread, we run into the issue that a dummy getsockopt or poll has to be done on the monitored socket for the event to be raised. This would be solved by doing this in the pipe class from the I/O thread, but the issue then becomes that the I/O thread, rather than just processing a command and forwarding it, has to acquire the monitor mutex, which is most likely way worse for performances.

I'd be inclined to leave this to be raised on the application thread. Thoughts? @somdoron @sigiesec

Reply:
How often does the mutex happen? Is it happening if monitoring is not enabled as well?

Reply:
No, only when the request is triggered and being processed - so once per probe

Reply:
how often a probe? is it by interval or user requesting it?

Reply:
user requesting it, although i'm considering allowing an interval as well, as it might be useful

Reply:
The monitoring needs a mutex, since the stored monitor socket is accessed from both the application (through zmq_socket_monitor*) and the I/O thread. Probably this could be solved without a mutex, but it has been like this for a long time.

Reply:
I considered waiting for the command, but especially without unit tests for the commands pipe it would get tricky avoiding regressions.